### PR TITLE
sb track2 iterators idea.

### PR DESCRIPTION
**Pull request by richardpark-msft**

### CODE REVIEW

File: sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+  iterator.close = async () => {
+    return new Promise((resolve) => {
+      receiver.drain = true;
+      receiver.addCredit(1);
+
+      receiver.once(ReceiverEvents.receiverDrained, () => {
+        resolve();
+      });
+    });
```
Comment (chradek) [inline sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts:21]: I do like this. What happens if we get any events while draining? I'm not sure how I'd get access to those (is it ridiculous to return any events received at this point? We're not closing the link so they could still be settled right?)

File: sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    receiver.addCredit(1);
+  }
+}
+
+interface PromisePayload {
+  promise: Promise<PromisePayload>;
+  message: string;
+}
+
+function setupReceiver(receiver: Receiver): Promise<PromisePayload> {
```
Comment (chradek) [inline sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts:53]: This is an interesting way to turn an event emitter into a chain of promises. I had to stare at it a while but this seems like it'd get the job done. 

I thought about what if we had a simpler implementation where you add a single credit, then resolve once you get a message. It'd be easier to reason about but then again this solution will let us handle prefetching.

Are you planning on having a buffer for messages like we have for event hubs? (useful for prefetching messages)

File: sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts (hunk 1 above)
Comment (richardpark-msft) [reply sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts:21]: It's not ridiculous - right now the semantics around drain are more unclear than I'd like. But I think what  you want to do make sense - just return what we have left. 

The only lame thing is that I've constrained the promise to only return a single event at a time. That could get fixed.

File: sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts (hunk 2 above)
Comment (richardpark-msft) [reply sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts:53]: I've had a few idle thoughts around this but I want to write them down. For instance, we could make the _user_ pass in a array (user being 'us', but from a higher level in the SDK). Then we just push onto an array and then the promise just signals that data has arrived.

(you had a similar/same idea with your Queue class)

File: sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts (hunk 1 above)
Comment (richardpark-msft) [reply sdk/servicebus/service-bus/src/receivers/bufferReceiver.ts:21]: Another thing that I've been thinking about is whether or thinking about 'drain' being the "final" event is correct or not. So we need to get confirmation on that as well (if so, returning the last few events here would be even more amazing)

**Pull request closed without merging**